### DISCUSSION TITLE
Version 4.0 Release

### DISCUSSION THREAD

**Discussion by GerryDazoo** [GerryDazoo/Slinger#529] [Announcements]
[V4.0_release_notes.pdf](https://github.com/GerryDazoo/Slinger/files/10723952/V4.0_release_notes.pdf)

And yes, Android and Firestick versions of Slinger Player are in the works.

**Comment by DVoorhees1957** [comment]
Thank you in advance for the forthcoming Android and Fire TV versions of Slinger Player!!

**Reply by GerryDazoo** [reply]
Take this for a test drive.
[SlingerPlayer.zip](https://github.com/GerryDazoo/Slinger/files/10733395/SlingerPlayer.zip)
I've only been able to test it with my Pixel 6. I guess, I'm going to have to write some documentation.

**Reply by tom76017** [reply]
Nice progress...  I tried a Firestick 4K and it froze.  I didn't think that the code would be that portable, but nice to see progress.
Perhaps a iOS iPad version someday???

**Reply by GerryDazoo** [reply]
That was the APK for an Android phone not the Firestick. Firestick is in progress due to the fact the navigation is totally different on the Stick.

I don't do apple. I have no way to test it.

**Reply by blasco403** [reply]
Hey Gerry,

I would "test" and report errors/issues for the Firestick version when at that stage.

I have 5 firesticks in Mexico, a 350 and 500. Slinger Server running on a PC in Alberta where the sling boxes are located. I have a always on vpn connecting my 2 networks, so can stream "local" to slinger server and also port forwarding for remote connection which is what I normally do for streaming.

I currently use Kodi on the Firestick as the client, works pretty good.

Would like access to a remote on the TV so I can use the air mouse like I used to with Slingplayer.

Looking forward to the "official" release no matter what.

Thanks for all the effort to keep them going.

Ron.

**Reply by RocketAlf** [reply]
Great! SlingerPlayer installed on ZTE Axon 10 and on 30. Runs fine. Many, many thanks Gerry.

**Reply by GerryDazoo** [reply]
Ron,
Yes, you'd need some additional hardware to properly navigate the Remote on Firestick. I don't have a plan to integrate the remote like I did for the android/phone version. I was just going to use the browser on my phone as a remote. Sorry for now. Maybe in the future when I better know what I'm going.

**Reply by RocketAlf** [reply]
Hello Gerry, I have now been able to spend a little longer with the Android SlingerPlayer. In the process, I have noticed the following. When I walk the dog, then I listen on the side in "audio only" mode via the Internet (mobile data 4G). During the 30 minutes or so of walking, I noticed several interruptions in the transmission (at about 5 minute intervals). I then restarted each time.
Previously, I also listened to the stream in "audio only" mode with VLC. There were also occasional interruptions of the transmission, but significantly less.
Is it possible to improve the stability of the transmission (e.g. by increasing time-out values or by automatically resuming the connection after an interruption)?

**Reply by GerryDazoo** [reply]
I'll see if I can reproduce your problem.

The code already automatically retries once if the stream drops. If it doesn't restart then it's because your data connection is poor. I routinely drive 30 minutes into the city and back streaming video the whole way with some minor interruptions but the code automatically recovers. 

Are you a using an Audio Only data stream from Slinger? If not that might explain a lot.

**Reply by RocketAlf** [reply]
Hi Gerry, it's me again, sorry about that. After 3 days of "test operation" of the SlingerPlayer app for Android outdoors (4G mobile data, always 4..5 of 6 bars in the field strength indicator, "audio only" mode), I must state that the transmission is very unstable. At intervals of 3 ... 5 minutes the connection breaks down and I have to restart manually.
At home in the WLAN, the transmission (with video) works without interference.  
On the same route outdoors (30 minutes dog walk) I have much less  or no interruptions with VLC. Something VLC must do differently (?).

Maybe the attached log files will help. I can't see any significant difference.
[SP_App-Log.txt](https://github.com/GerryDazoo/Slinger/files/10775387/SP_App-Log.txt)
[VLC-log.txt](https://github.com/GerryDazoo/Slinger/files/10775388/VLC-log.txt)

**Comment by mdehrlich** [comment]
Gerry,
I have v4 running.  I was able to add my three lines of code to support my special remote features.  Thanks for your efforts incorporating suggestions.  
I haven't tried slingerplayer yet.
Reminder:  my special feature is a drop down menu to select root boxnames.  Mods in slinger keep track of the current stream and route the remote requests to the specific remote.  No impact to normal use.
I also added one more line to support RESET in my drop down menu.  When Slinger gets /Remote/RESET it kills itself.  This way I do not need to add a reset key.
mike

**Reply by GerryDazoo** [reply]
With the 4.0 API you can do the reset without having to mod the code.
Send this to your [SERVER] port.
SENDIRKEYCODES=-1

**Reply by mdehrlich** [reply]
I don't know how to simulate a button action.
Groping in the dark on this.  Do I send /Remote/MainHD?SENDIRKEYCODES=-1  I think I can do this.

**Reply by mdehrlich** [reply]
looking at V4 code, looks like I send ip:port/SENDIRKEYCODES=-1
I'll try it.

**Reply by mdehrlich** [reply]
I tried everything I can think of.  Cannot make it work.

**Reply by GerryDazoo** [reply]
md; It's not a web address it's just a TCP packet. You send SENDIRKEYCODES=-1 directly to the server ip and port. (No / ), You can do it from a linux command line with this little snippet
echo -n "SENDIRKEYCODES=-1" | nc 192.168.117.122 65432
not sure what the windows equivalent would be.
So simply, open a TCP socket to ip:port, send SENDIRKEYCODES=-1, close socket.

You can also do it with a http POST request (not GET).
curl -X POST http://ip:port/Remote/MainHD -d Restart=-1

**Reply by mdehrlich** [reply]
To do it in my remote, I need javascript or html?

**Reply by GerryDazoo** [reply]
With javascript you can send a POST with the "fetch" api call. I'm not an expert but I suspect it would be possible to send a raw TCP packet from Javascript as well.

**Reply by mdehrlich** [reply]
I was able to make it work, I learned a lot from this project.
Two suggestions to modify v4.  They are not required, just clear up some possible confusion.
1) when looking for /Remote, limit the search for 'remote' to the first 20 characters:   if ( 'remote' in data[0:20].lower() ........
2) right after that when looking for a stream start ' elif 'GET'  and 'HTTP' in data:',  you really want is 'elif 'GET' in data and 'HTTP' in data:'

Here is what I did in javascript for my remote
let xhr = new XMLHttpRequest();
xhr.open("POST", document.URL);
xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
if (confirm("Restart Slinger?")) {
            xhr.send("Restart=-1") ;
} else { 
            alert("Restart Cancelled"); 
}

**Reply by mdehrlich** [reply]
This programing reminds me of my old assembler coding days.  We had a joke that all programming was going to become a single line of code called 'Do_it'  The parameters are going to kill you!

**Reply by mdehrlich** [reply]
This also works: (Is it better?)
 if (box == "RESTART") { 
                                        if (confirm("Restart Slinger?")) {
                                                fetch(document.URL, {
                                                method: 'POST',
                                                headers: {'Content-type': 'application/x-www-form-urlencoded'},
                                                body: "Restart=-1"
                                                })
                                        } else { 
                                                alert("Restart Cancelled"); 
                                        }                          
 } else {

**Comment by remgr** [comment]
Hi Gerry,
Tried this on my Pixel 6, works perfectly.  Initially has a bit of a video stutter, but smoothed out after a while.  Loved the remote on screen rotation.
As always, thanks for your awesome work.

**Comment by socelillyhammer** [comment]
Read the release notes, but have yet to install it.  Thank you!
Would also like to say thanks for everything.  I have the slinger working while im currently overseas and using the mx player on the firestick, it's truly amazing what you all have colletively built.   So a big thank you to all of you guys again.

**Reply by bradut** [reply]
Two months after release I've tried the Server v4.01 and the SlingerPlayer, and they work great. 
Thank you for everything Gerry !
